File: ckl-dotnet/ckl-dotnet/functions/FuncExecute.cs
/*  Copyright (c) 2021 Damian Brunold, Gesundheitsdirektion Kanton Zürich

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
*/

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;

namespace CheckerLang
{
    public class FuncExecute : FuncBase
    {
        public FuncExecute() : base("execute")
        {
             info = "execute(program, args, work_dir = NULL, echo = FALSE, output_file = NULL)\r\n" +
                 "\r\n" +
                 "Executed the program and provides the specified arguments in the list args.\r\n";
        }

        public override bool IsSecure()
        {
            return false;
        }
        
        public override List<string> GetArgNames()
        {
            return new List<string> {"program", "args", "work_dir", "echo", "output_file"};
        }
        
        public override Value Execute(Args args, Environment environment, SourcePos pos)
        {
            var program = args.GetString("program").GetValue();
            var arguments = args.GetList("args").GetValue();
            string work_dir = null;
            if (args.HasArg("work_dir")) work_dir = args.GetString("work_dir").GetValue();
            var echo = false;
            if (args.HasArg("echo")) echo = args.GetBoolean("echo").GetValue();
            string output_file = null;
            if (args.HasArg("output_file")) output_file = args.GetString("output_file").GetValue();
            try
            {
                var list = new StringBuilder();
                foreach (var argument in arguments) {
                    list.Append(argument.AsString().GetValue()).Append(" ");
                }
                list.Remove(list.Length - 1, 1);
                var process = new Process();
                var info = process.StartInfo;
                info.FileName = program;
                info.Arguments = list.ToString();
                if (work_dir != null) info.WorkingDirectory = work_dir;
                info.CreateNoWindow = true;
                info.RedirectStandardOutput = true;
                info.RedirectStandardError = false;
                info.RedirectStandardInput = false;
                if (echo) System.Console.WriteLine(program + " " + list);
                process.Start();
                var output = process.StandardOutput.ReadToEnd();
                if (output_file != null)
                {
                    File.WriteAllText(output_file, output);
                }
                else
                {
                    Console.Write(output);
                }
                process.WaitForExit();
                return new ValueInt(process.ExitCode);
            } catch (Exception e) {
                throw new ControlErrorException(new ValueString("ERROR"), "Cannot execute " + program + ": " + e.Message, pos);
            }
         }
    }
    
}

File: ckl-java/src/main/ch/checkerlang/functions/FuncExecute.java
/*  Copyright (c) 2021 Damian Brunold, Gesundheitsdirektion Kanton Zürich

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE.
*/
package ch.checkerlang.functions;

import ch.checkerlang.Args;
import ch.checkerlang.ControlErrorException;
import ch.checkerlang.Environment;
import ch.checkerlang.SourcePos;
import ch.checkerlang.values.Value;
import ch.checkerlang.values.ValueInt;
import ch.checkerlang.values.ValueNull;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FuncExecute extends FuncBase {
    public FuncExecute() {
        super("execute");
        info = "execute(program, args, work_dir = NULL, echo = FALSE)\r\n" +
                "\r\n" +
                "Executed the program and provides the specified arguments in the list args.\r\n";
    }

    public boolean isSecure() {
        return false;
    }

    public List<String> getArgNames() {
        return Arrays.asList("program", "args", "work_dir", "echo");
    }

    public Value execute(Args args, Environment environment, SourcePos pos) {
        String program = args.getString("program").getValue();
        List<Value> arguments = args.getList("args").getValue();
        String work_dir = null;
        if (args.hasArg("work_dir")) work_dir = args.getString("work_dir").getValue();
        boolean echo = false;
        if (args.hasArg("echo")) echo = args.getBoolean("echo").getValue();
        try {
            List<String> list = new ArrayList<>();
            list.add(program);
            for (Value argument : arguments) {
                list.add(argument.asString().getValue());
            }
            ProcessBuilder proc = new ProcessBuilder(list);
            proc.inheritIO();
            proc.directory(work_dir == null ? null : new File(work_dir));
            if (echo) System.out.println(proc.command());
            Process process = proc.start();
            process.waitFor();
            return new ValueInt(process.exitValue());
        } catch (Exception e) {
            throw new ControlErrorException("Cannot execute " + program + ": " + e.getMessage(), pos);
        }
    }

}
